File: Algorithm_299_Bulls and Cows.py
# 思路：
# 遍历两字符串，数组secretCount和guessCount中记录当前数字出现的次数
# 若当前位置上的字符相等，则bulls += 1，不记录在两数组中
# 否则，数组secretCount和guessCount对应位置+1
# 最后，再遍历一遍两数组，在各个位置上都取较小值加到cows上
# 较小值代表当前数字出现在两字符串中出现次数较少的哪一个

class Solution:
    def getHint(self, secret: str, guess: str) -> str:
        secretCount = [0 for _ in range(10)]
        guessCount = [0 for _ in range(10)]
        bulls = cows = 0
        for i in range(len(secret)):
            n1 = int(secret[i])
            n2 = int(guess[i])
            if n1 == n2:
                bulls += 1
            else:
                secretCount[n1] += 1
                guessCount[n2] += 1
        for i in range(10):
            cows += min(secretCount[i], guessCount[i])
        return str(bulls) + 'A' + str(cows) + 'B'
    

    
class SolutionOptimized:
    def getHint(self, secret: str, guess: str) -> str:
        numsCount = [0 for _ in range(10)]
        bulls = cows = 0
        for i in range(len(secret)):
            n1 = int(secret[i])
            n2 = int(guess[i])
            if n1 == n2:
                bulls += 1
            else:
                if numsCount[n1] < 0:
                    cows += 1
                if numsCount[n2] > 0:
                    cows += 1
                numsCount[n1] += 1
                numsCount[n2] -= 1
        return str(bulls) + 'A' + str(cows) + 'B'

File: Algorithm_299_Bulls and Cows.java
// 方法：用数组digitCount记录下secret中0-9每个数字出现的次数
// 然后遍历guess两遍。第一遍里，找到所有的bull，并相应地从digitCount中减去各数字出现的次数
// 第二遍找到所有的cow
// 注意这两遍遍历不能合并，不然有可能出现cow把digitCount[guessedDigit]减到0，导致bull结果错误的情况

class Solution {
    public String getHint(String secret, String guess) {
        int n = secret.length();
        int[] digitCount = new int[10];
        for (int i = 0; i < n; i++) {
            int digit = secret.charAt(i) - '0';
            digitCount[digit]++;
        }

        int bull = 0, cow = 0;
        for (int i = 0; i < n; i++) {
            int guessedDigit = guess.charAt(i) - '0';
            int correctDigit = secret.charAt(i) - '0';
            if (guessedDigit == correctDigit) {
                bull++;
                digitCount[guessedDigit]--;
            }
        }

        for (int i = 0; i < n; i++) {
            int guessedDigit = guess.charAt(i) - '0';
            int correctDigit = secret.charAt(i) - '0';
            if (guessedDigit != correctDigit && digitCount[guessedDigit] > 0) {
                cow++;
                digitCount[guessedDigit]--;
            }
        }

        return bull + "A" + cow + "B";
    }
}